File: android/src/main/kotlin/com/giphy/giphy_flutter_sdk/GiphyFlutterSdkPlugin.kt
package com.giphy.giphy_flutter_sdk

import android.content.Context
import android.app.Activity
import androidx.fragment.app.FragmentActivity
import com.giphy.giphy_flutter_sdk.utils.require
import com.giphy.giphy_flutter_sdk.utils.initializeVideoCache
import com.giphy.giphy_flutter_sdk.utils.getVideoPlayerFactory
import com.giphy.giphy_flutter_sdk.dto.GiphyFlutterTheme
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

import com.giphy.sdk.ui.Giphy
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding

class GiphyFlutterSdkPlugin : FlutterPlugin, MethodChannel.MethodCallHandler, ActivityAware {
    private lateinit var context: Context
    private lateinit var channel: MethodChannel
    private var activity: Activity? = null
    private val giphyDialogHandler = GiphyFlutterDialog()
    private val giphyVideoManagerHandler = GiphyFlutterVideoManager()

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "com.giphyfluttersdk")
        channel.setMethodCallHandler(this)
        giphyDialogHandler.onAttachedToEngine(flutterPluginBinding)
        giphyVideoManagerHandler.onAttachedToEngine(flutterPluginBinding)

        flutterPluginBinding
            .platformViewRegistry
            .registerViewFactory(
                "com.giphyfluttersdk/mediaView",
                GiphyFlutterMediaViewFactory(flutterPluginBinding.binaryMessenger)
            )

        flutterPluginBinding
            .platformViewRegistry
            .registerViewFactory(
                "com.giphyfluttersdk/videoView",
                GiphyFlutterVideoViewFactory(flutterPluginBinding.binaryMessenger)
            )

        flutterPluginBinding
            .platformViewRegistry
            .registerViewFactory(
                "com.giphyfluttersdk/gridView",
                GiphyFlutterGridViewFactory(flutterPluginBinding.binaryMessenger)
            )
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        if (call.method == "configure") {
            val apiKey = call.argument<String>("apiKey").require("API key is missing")
            val verificationMode = call.argument<Boolean>("verificationMode") ?: false
            val videoCacheMaxBytes = call.argument<Int>("videoCacheMaxBytes") ?: 100 * 1024 * 1024
            configureGiphy(apiKey, verificationMode, videoCacheMaxBytes)
            result.success(null)
        } else {
            result.notImplemented()
        }
    }

    private fun configureGiphy(apiKey: String, verificationMode: Boolean, videoCacheMaxBytes: Int) {
        initializeVideoCache(
            context, videoCacheMaxBytes.toLong()
        )
        val player = getVideoPlayerFactory()
        Giphy.videoPlayer = player
        val appInfo = GiphyFlutterSdkInfo(context)
        Giphy.configure(
            context,
            apiKey,
            verificationMode,
            metadata = hashMapOf(appInfo.name to appInfo.version)
        )
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        giphyDialogHandler.onDetachedFromEngine(binding)
        giphyVideoManagerHandler.onDetachedFromEngine(binding)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        giphyDialogHandler.setActivity(activity as? FragmentActivity)
        GiphyFlutterTheme.getInstance().setActivity(activity)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
        giphyDialogHandler.setActivity(null)
        GiphyFlutterTheme.getInstance().setActivity(null)
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
        giphyDialogHandler.setActivity(activity as? FragmentActivity)
        GiphyFlutterTheme.getInstance().setActivity(activity)
    }

    override fun onDetachedFromActivity() {
        activity = null
        giphyDialogHandler.setActivity(null)
        GiphyFlutterTheme.getInstance().setActivity(null)
    }
}

File: ios/Classes/GiphyFlutterSdkPlugin.swift
import Flutter
import UIKit
import GiphyUISDK

public class GiphyFlutterSdkPlugin: NSObject, FlutterPlugin {
    
    var giphyDialogHandler: GiphyFlutterDialog?
    var giphyVideoManagerHandler: GiphyFlutterVideoManager?
    private var channel: FlutterMethodChannel?
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "com.giphyfluttersdk", binaryMessenger: registrar.messenger())
        let instance = GiphyFlutterSdkPlugin()
        instance.channel = channel
        registrar.addMethodCallDelegate(instance, channel: channel)
        
        instance.giphyDialogHandler = GiphyFlutterDialog()
        instance.giphyDialogHandler?.onAttachedToEngine(flutterPluginBinding: registrar)
        
        instance.giphyVideoManagerHandler = GiphyFlutterVideoManager()
        instance.giphyVideoManagerHandler?.onAttachedToEngine(flutterPluginBinding: registrar)
        
        let mediaViewFactory = GiphyFlutterMediaViewFactory(messenger: registrar.messenger())
        registrar.register(mediaViewFactory, withId: "com.giphyfluttersdk/mediaView")
        
        let videoViewFactory = GiphyFlutterVideoViewFactory(messenger: registrar.messenger())
        registrar.register(videoViewFactory, withId: "com.giphyfluttersdk/videoView")
        
        let gridViewFactory = GiphyFlutterGridViewFactory(messenger: registrar.messenger())
        registrar.register(gridViewFactory, withId: "com.giphyfluttersdk/gridView")
    }
    
    public func detachFromEngine(for registrar: FlutterPluginRegistrar) {
        channel?.setMethodCallHandler(nil)
        
        giphyDialogHandler?.onDettachedFromEngine(flutterPluginBinding: registrar)
        giphyDialogHandler = nil
        
        giphyVideoManagerHandler?.onDettachedFromEngine(flutterPluginBinding: registrar)
        giphyVideoManagerHandler = nil
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "configure":
            guard let args = call.arguments as? [String: Any?] else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments received", details: nil))
                return
            }
            
            let apiKey: String = args["apiKey"] as? String ?? ""
            if apiKey.isEmpty {
                result(FlutterError(code: "MISSING_API_KEY", message: "API key is missing", details: nil))
                return
            }
            
            let verificationMode: Bool = args["verificationMode"] as? Bool ?? false
            let videoCacheMaxBytes: Int = args["videoCacheMaxBytes"] as? Int ?? 0
                        
            configureGiphy(apiKey: apiKey, verificationMode: verificationMode, videoCacheMaxBytes: videoCacheMaxBytes)
            
            result(nil)
            
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func configureGiphy(apiKey: String, verificationMode: Bool, videoCacheMaxBytes: Int) {
        let metadata = [GiphyFlutterSdkInfo.shared.name: GiphyFlutterSdkInfo.shared.version ?? ""]
        Giphy.configure(
            apiKey: apiKey as String,
            verificationMode: verificationMode,
            metadata: metadata
        )
    }
}
